Metadata for Requiem

ProbablyManuel opened the pull request:
In short: This is all the metadata needed to create a good Requiem-based load order.

It points out known incompatibilities, mods that require a patch, redundant plugins and ensures a correct load order for Requiem and its patches. Explaining every single rule would take way too long and  it almost always boils down to "it's recommended by the Requiem Dungeon Masters (the developers)" or "it's recommended by the patch maker". I hope it suffices that I'm part of the Requiem Dungeon Masters and one of the major patch makers for Requiem and that this metadata has been tested for three months.
I compiled the metadata in a userlist.yaml and uploaded it in March on my [patch central](http://www.nexusmods.com/skyrim/mods/61621/?) for public testing. The last version had over 200 unique downloads. When no more issues were reported for several weeks I moved the metadata to the masterlist and put it up for testing again in case I overlooked something during the process and didn’t catch it in local testing. I got a few positive and no negative reports so I guess it’s now ready for merge.

Additional metadata that isn't strictly related to Requiem but is required for a correct load order with Requiem (and of course without Requiem too):
- [ELE](http://www.nexusmods.com/skyrim/mods/59733/?) should be loaded after all plugins that aren't dynamically created like e.g. DynDOLOD because any additional or edited reference also adds a (necessary) ITM record of the parent cell/worldspace to the plugin, that then overwrites ELE. I set its global priority to 77.
- [No Spell Tomes as Loot or From Vendors](http://www.nexusmods.com/skyrim/mods/81341/?) No mod **ever** has any reason to overwrite it. Dynamically created plugins like a bashed patch may even undo the changes. I set its global priority to 105.
- [Skyrim Unlocked](http://www.nexusmods.com/skyrim/mods/69420/?) should be loaded at the bottom of the load order like LAL or Open Cities. I set its global priority to 70.

ProbablyManuel committed:
Added Requiem.esp

ProbablyManuel committed:
Replaced oudated patch rules

ProbablyManuel committed:
Added new patch rules

ProbablyManuel committed:
Added templates

ProbablyManuel committed:
Added mods that go after Requiem

ProbablyManuel committed:
Fixed duplicate EBQO entry

ProbablyManuel committed:
Fixed duplicate entry of ELE

ProbablyManuel committed:
Added warnings about missing Requiem patches

ProbablyManuel committed:
Fixed error at line 2524

ProbablyManuel committed:
Fixed error at line 7563

ProbablyManuel committed:
Fixed error at line 16436

ProbablyManuel committed:
Fixed error at line 18245

ProbablyManuel committed:
Added warnings about redundant plugins

ProbablyManuel committed:
Improved warning text

ProbablyManuel committed:
Change global priority of ELE to 77

ProbablyManuel committed:
Minor fixes

Increased global priority of No Vendor or Loot Spells.esp
Removed a debug message
Moved Reqtificator to other SkyProc patcher

ProbablyManuel committed:
Lowered global priority of Skyrim Unlocked to 70

ProbablyManuel committed:
Improved documentation

ProbablyManuel committed:
Added redundancy warning for Height Adjusted Races.esp

ProbablyManuel committed:
Fixed redundancy warning about RUSTIC SOULGEMS - Unsorted.esp not appearing

Freso commented on masterlist.yaml:
```diff
@@ -22544,8 +22850,6 @@ plugins:
   - name: 'Alternate Start - Live Another Life.esp'
     global_priority: 70
-    tag:
-      - C.Location
```
Why are you removing this tag?

Ortham replied:
As above, it got moved down a bit for some reason.

Freso replied:
Ah, right. I was sure I had checked and double checked for this. Apparently I should have triple checked too. :)

ProbablyManuel replied:
Same reason as in [#136 (comment)](https://github.com/loot/skyrim/pull/136#discussion_r122197248)

Freso commented on masterlist.yaml:
```diff
@@ -17632,12 +17910,15 @@ plugins:
       - 'Atlas Map Markers.esp'
       - 'The Horkerpocalypse.esp'
-    tag:
-      - C.Location
```
?

Ortham replied:
It's been moved to below the `msg` block. I'd rather it wasn't, it's just noise, but I don't consider it a blocker.

Freso replied:
Ah, I missed that. I thought it was removed entirely like for https://github.com/loot/skyrim/pull/136#discussion_r122011799

ProbablyManuel replied:
I followed the order [here](https://loot.github.io/docs/0.9.2/LOOT%20Metadata%20Syntax.html) that seems to be used by the majority of the masterlist.

Freso commented on masterlist.yaml:
```diff
@@ -17921,6 +18202,7 @@ plugins:
     after:
       - 'Immersive Citizens - AI Overhaul.esp'
   - name: 'ELE_Legendary_Lite.esp'
+    global_priority: 77
```
What were your thoughts on picking this specific `global_priority`?

(Also, https://loot.github.io/docs/contributing/Global-Priorities.html really ought to get updated. :/)

Ortham replied:
> (Also, https://loot.github.io/docs/contributing/Global-Priorities.html really ought to get updated. :/)

I thought I'd already done this when I standardised priority values across the masterlists... I'll have to dig out the forum post I made about it.

ProbablyManuel replied:
Higher than Requiem but lower than dynamic plugins.

Freso commented on masterlist.yaml:
```diff
@@ -24276,19 +24579,19 @@ plugins:
       - 'Convenient Horses.esp'
-
   - name: 'Undeath.esp'
-    url: [ 'http://www.nexusmods.com/skyrim/mods/40607' ]
```
Why are you removing this URL? It still works fine.

Ortham replied:
It got moved below the `msg` block, I'd rather it didn't.

Freso replied:
Darn. These plugin entries are 2busy4me!

ProbablyManuel replied:
Same reason as in [#136 (comment)](https://github.com/loot/skyrim/pull/136#discussion_r122197248)

Freso commented on masterlist.yaml:
```diff
@@ -2501,16 +2523,49 @@ plugins:
         udr: 0
+  - name: 'AOS2_DD Patch.esp'
+    msg:
+      - <<: *redundantPluginX
```
Why not `*alreadyInX`? Seems like a perfect fit.

ProbablyManuel replied:
Same reason as in [#136 (comment)](https://github.com/loot/skyrim/pull/136#discussion_r122193102)

Freso commented on masterlist.yaml:
```diff
@@ -24664,3 +24973,276 @@ plugins:
+    msg:
+      - <<: *reqRequiemPatch
+        condition: 'file("Requiem.esp") and not active("Requiem - 83Willows 101Bugs.esp")'
+      - <<: *redundantPluginX
```
Wouldn't `*useOnlyOneX` be appropriate here?

ProbablyManuel replied:
`Requiem - 83Willows 101Bugs.esp` is both a patch and a replacer plugin (but with a different name than the original plugin). The rule is:
Do not use `83Willows_101BUGS_V4_HighRes_HighSpawn.esp` if `Requiem - 83Willows 101Bugs.esp` is installed.
and not:
Use only one of `Requiem - 83Willows 101Bugs.esp` and `83Willows_101BUGS_V4_HighRes_HighSpawn.esp`.

Freso commented on masterlist.yaml:
```diff
@@ -188,6 +188,11 @@ common:
+    type: warn
+    content:
+      - lang: en
+        text: 'This plugin is redundant when using %1% and should be deactivated or deleted.'
```
Looking over the uses of this, it seems they would all be appropriate for either `&alreadyInX` or `&useOnlyOneX`. Are there any you feel wouldn't fit into either of these?

ProbablyManuel replied:
I dislike the imperative delete in `&alreadyInX`. Disabling is just as good and it's the safer option for the average user. However for e.g. `&AlreadyInUSKP` you must delete the mod and not just disable the plugin or else there might be weird conflicts with outdated loose files overwriting the USKP.
`&useOnlyOneX` fits well for some of these entries and is more precise. I'll change that.

ProbablyManuel replied:
I've been thinking about this some more:
`&alreadyInX` is fine for cases like 'AOS2_DD Patch.esp'. It's clear that you must delete this patch and not the whole mod and deleting is more appropriate than disabling because these plugins should never have been installed in the first place. The problem are cases like 'RUSTIC SOULGEMS - Unsorted.esp', where the patch is also a replacer plugin (but uses a different name). Here you must disable/delete only the plugin but keep the meshes and textures which is the opposite of what's required in `&AlreadyInUSKP`.

Would it resolve your concern if I used your suggestion for the former case and used a new replacer plugin message for the latter?

Ortham replied:
A new message specific to the case where resources should be kept seems more appropriate to me than using `alreadyInX`.

Freso commented on masterlist.yaml:
```diff
@@ -228,11 +233,16 @@ common:
     subs: [ '[Better males - Beautiful nudes and faces - New hairstyles](http://www.nexusmods.com/skyrim/mods/2488)' ]
-
-  - &reqSkyUI # Useful for when a plugin needs SkyUI, but shouldn't load after it (eg. when it's a master plugin).
+  # Useful for when a plugin needs SkyUI, but shouldn't load after it (eg. when it's a master plugin).
```
Any non-cosmetic reason for this move? (I don't disagree with the move! Just curious. :))

ProbablyManuel replied:
No, it just bothered me because it blatantly ignored this
> Make as many silent comments as you want within the masterlists - it doesn't hurt (within reason). However, make sure to put them _above_ the line(s) that they refer to.

Freso commented on masterlist.yaml:
```diff
@@ -228,11 +233,16 @@ common:
+    type: warn
+    content:
+      - lang: en
+        text: 'Incompatible with: Requiem. You must install a compatibility patch.'
```
I don't think this message is very helpful without a reference to what compatibility patch to use and/or where to get it. (Maybe it is though. I have no experience at all with Requiem.)

ProbablyManuel replied:
I agree that it's not very helpful but that's somewhat the intention. Requiem patches are not a straight forward record patching and more about balancing. For some mods there are multiple patches by different authors or patches for different versions of Requiem. Sometimes linking to a specific patch may be misleading because said patch is slightly outdated or does something some users consider gameplay or immersion-breaking and for them it might be better not to use the mod at all.
I believe a generic message that points out incompatibility is better than linking all possible patches. The Requiem community is very helpful and google is a powerful friend. And it's far easier to maintain in the masterlist ;)

Ortham replied:
Is there a site / page that hosts information on what the latest patches are, etc. that you could link to instead of specific patches?

Freso replied:
No objections to this based on your comment, but I would like to know the answer to @WrinklyNinja's question. :)

ProbablyManuel replied:
Not really 😅 The Requiem Dungeon Masters once attempted to set up a user-driven page that hosts information on what the latest patches are, etc but there wasn't any real interest so it got suspended after a while.

Freso replied:
What a shame. C'est la vie though.

Freso commented on masterlist.yaml:
```diff
@@ -8628,19 +8716,37 @@ plugins:
       - Delev
       - Relev
+    url:
+      - 'http://www.nexusmods.com/skyrim/mods/19281/?'
```
I personally prefer to have the `url` right under the `name`, to make it easy to look up. Also note how (most, at least) other `url` entries do not have the trailing "/?". :)

ProbablyManuel replied:
I followed the order [here](https://loot.github.io/docs/0.9.2/LOOT%20Metadata%20Syntax.html) that seems to be used by the majority of the masterlist.

Freso commented on masterlist.yaml:
```diff
@@ -8654,16 +8760,12 @@ plugins:
-      - crc: 0xa3a78f22
-        <<: *dirtyPlugin
-        itm: 1
+  # Must be loaded directly after Requiem.esp
```
This should probably be codified rather than stated in a comment.

Ortham replied:
It's codified as best as possible in the priority value, isn't it? We don't have a better way of saying "load directly after".

Freso commented on masterlist.yaml:
```diff
@@ -8654,16 +8760,12 @@ plugins:
+  - name: 'Requiem - Bugsmasher Edition.esp'
+    priority: -127
+  # Must be loaded directly after Requiem.esp
+  - name: 'Requiem - Legendary Bugsmasher Edition.esp'
```
Are this and `Requiem - Bugsmasher Edition.esp` `inc`ompatible?

ProbablyManuel replied:
Technically no but it requires you to have both USLEEP and USKP active, that are of course incompatible (and already in the masterlist).

Freso commented on masterlist.yaml:
```diff
@@ -8680,37 +8782,131 @@ plugins:
+    req:
+      - 'Requiem.esp'
+  # other mods that must be loaded after Requiem but are listed somewhere else in LOOT's rules:
+  # - RealisticWaterTwo.esp
```
The `after` specifies that they must load after `Requiem.esp`. No need to make a comment like this about it. It's just clutter.

ProbablyManuel replied:
I added that comment so that all mods that must be loaded after Requiem are listed in one section as there is no way to search for `after` `Requiem.esp`. If you really think that's just clutter I'll remove it.

Ortham replied:
I think it's best to remove the comment, for the additional reason that it's not necessarily an exhaustive list, if new `after` metadata is added that's Requiem-related, the comment will become out of date (and it's hard enough to keep comments up to date when they sit even inline with what they're commenting).

Freso commented on masterlist.yaml:
```diff
@@ -8628,19 +8716,37 @@ plugins:
     msg: [ *obsolete ]
+### Rules for Requiem and its patches ###
   - name: 'Requiem.esp'
+    global_priority: 75
```
That's a **heavy** `global_priority`. Why that number?

ProbablyManuel replied:
The rule of thumb is:
> Requiem should be the last mod in your load order to win potential conflicts with other mods. Only load other mods after Requiem, if you have a good reason to do so, like an explicit advice for Requiem from the author of the other mod.

It prevents all kinds of unintended side-effects, distorted gameplay and false bug reports.

Looking once more at other mods with a high `global_priority`, I think it could be reduced to 70 because all other mods with that priority require a Requiem patch anyway.

Freso reviewed (changes requested):
I mostly have a number of questions for you, but I also some smaller nitpicks. Thanks for your efforts in this so far though! I hope you will stick with it just a tad longer. :)

Ortham commented:
Well, I don't know much about Requiem, but it generally looks good to me.

*Edit: Looks like Freso's been paying more attention than me...*

ProbablyManuel committed:
Replaced redundantPluginX with useOnlyOneX

Replaced redundantPluginX with useOnlyOneX if appropriate
Removed trailing "/?"

ProbablyManuel committed:
Fixed error at line 25199

ProbablyManuel committed:
Final edits

Added &replacerPluginForX template
Replaced &redundantPluginX by &alreadyInX and &replacerPluginForX
Lowered global_priority of Requiem.esp to 70
Lowered global_priority of ELE to 75
Removed clutter comment

ProbablyManuel committed:
Fixed error at line 8906

ProbablyManuel committed:
Fixed error at line 8931

Freso commented on masterlist.yaml:
```diff
@@ -2501,16 +2530,49 @@ plugins:
   - name: 'AOS.esp'
     after:
       - 'WetandCold.esp'
+      # RWT must be loaded after Requiem but AOS shouldn't so this rule is disabled when Requiem is installed
```
But this rule (conditionally) says to load `AOS.esp` `after` `RealisticWaterTwo.esp`, not after Requiem... ? Does RWT load after Requiem or did you make a typo or... ?

ProbablyManuel replied:
RWT must be loaded after Requiem and AOS must be loaded after RWT unless the AOS-RWT patch is installed. These rules are in the masterlist. Logically if you use all three mods together AOS is loaded after Requiem but it shouldn't. In that case AOS should fall behind RWT and Requiem and the patch should be used.

Freso reviewed (commented):
I think the YAML anchors look neater now, using more specific/targeted comments rather than all using a blanket one. I only have two pending questions now, I think. :)

Ortham reviewed (changes requested):
LGTM, pending an answer to my question about the generic link.

Freso commented:
> LGTM, pending an answer to my question about the generic link.

Oh, crud, I didn't see your comment here. Sorry. :( Are you referring to https://github.com/loot/skyrim/pull/136#discussion_r122292339 ?

Ortham commented:
> Oh, crud, I didn't see your comment here. Sorry. :( Are you referring to #136 (comment) ?

Yes, for some reason the comments after didn't display for me. :/ LGTM, anyway.
